aten/src/ATen/native/cuda/TensorModeKernel.cu

```diff
@@ -69,11 +83,11 @@ void calculate_mode(
                    thrust::not_equal_to<scalar_t>());
 
   // Count frequency of each element
-  Tensor keys = at::empty(unique, self.options());
-  Tensor counts = at::empty(unique, self.options().dtype(kLong));
+  auto keys = cuda_allocator->allocate(unique * sizeof(scalar_t));
+  auto counts = cuda_allocator->allocate(unique * sizeof(int64_t));
 
-  auto keys_ptr = thrust::device_pointer_cast(keys.data_ptr<scalar_t>());
-  auto counts_ptr = thrust::device_pointer_cast(counts.data_ptr<int64_t>());
+  auto keys_ptr = thrust::device_pointer_cast(static_cast<scalar_t*>(keys.get()));
```

Comment:
there's a common pattern here used at least three times: cuda allocate N elements of type T. This should probably be a utility to avoid the duplicating of the type and the potential to get that wrong.

The wrench is that the storage's lifetime is managed by result from the allocator.

What roughly makes sense to me:
```
template <typename T>
class CudaArray {
 public:
  CudaArray(Allocator& allocator, std::size_t size) : CudaArray(allocator.allocate(size * sizeof(T))) {}
 
  T* as_ptr() { return ptr_; }
  // ... blah blah
 private:
  explicit CudaAllocator(DataPtr data) : data_(std::move(data)), ptr_(thrust::device_pointer_cast(static_cast<T*>(data_.get()))) {}
  DataPtr data_;
  T* ptr_;
};

Reply:
thanks, i like the way you did it better.